src/ip/Address.cc

```diff
@@ -683,32 +677,73 @@ Ip::Address::InitAddr(struct addrinfo *&ai)
     }
 
     // remove any existing data.
-    if (ai->ai_addr) delete ai->ai_addr;
-
-    ai->ai_addr = (struct sockaddr*)new sockaddr_in6;
-    memset(ai->ai_addr, 0, sizeof(struct sockaddr_in6));
-
-    ai->ai_addrlen = sizeof(struct sockaddr_in6);
+    FreeAddrMember(*ai);
 
+    AllocateAddrMember<struct sockaddr_in6>(*ai);
 }
 
 void
 Ip::Address::FreeAddr(struct addrinfo *&ai)
 {
     if (ai == nullptr) return;
 
-    if (ai->ai_addr) delete ai->ai_addr;
-
-    ai->ai_addr = nullptr;
-
-    ai->ai_addrlen = 0;
+    FreeAddrMember(*ai);
 
     // NP: name fields are NOT allocated at present.
     delete ai;
 
     ai = nullptr;
 }
 
+/// Allocates and fills with zeros a SockAddrType object.
+/// Initializes addrinfo::ai_addr and ai_addrlen with the allocated object.
+/// \returns the allocated object
+template <class SockAddrType>
+SockAddrType &
+Ip::AllocateAddrMember(struct addrinfo &ai)
+{
+    static_assert(
+        std::is_same<SockAddrType, struct sockaddr_in>::value ||
+        std::is_same<SockAddrType, struct sockaddr_in6>::value,
+        "FreeAddrMember() supports this addrinfo::ai_addr type");
+
+    const auto ai_addr = new SockAddrType;
+    // We do not use `new SockAddrType{}` above instead of memset() below
+    // because, since C++14, doing so may not initialize SockAddrType padding.
+    static_assert(std::is_trivial<SockAddrType>::value, "can memset() this addrinfo::ai_addr type");
+    memset(ai_addr, 0, sizeof(*ai_addr));
+
+    ai.ai_addr = reinterpret_cast<struct sockaddr*>(ai_addr);
+    ai.ai_addrlen = sizeof(*ai_addr);
+
+    return *ai_addr;
+}
+
+/// Deallocates addrinfo::ai_addr member of the given structure and adjusts that
+/// structure accordingly. Safe for structures with a nil ai_addr member.
+/// \sa Ip::AllocateAddrMember()
+void
+Ip::FreeAddrMember(struct addrinfo &ai)
+{
+    if (!ai.ai_addr) {
+        return;
+    }
+
+    if (ai.ai_addrlen == sizeof(struct sockaddr_in)) {
+        delete reinterpret_cast<struct sockaddr_in*>(ai.ai_addr);
+    } else if (ai.ai_addrlen == sizeof(struct sockaddr_in6)) {
+        delete reinterpret_cast<struct sockaddr_in6*>(ai.ai_addr);
```

Comment:
Current style guidelines is not to add braces around single-line if-statements.
```suggestion
    if (ai.ai_addrlen == sizeof(struct sockaddr_in))
        delete reinterpret_cast<struct sockaddr_in*>(ai.ai_addr);
        
    else if (ai.ai_addrlen == sizeof(struct sockaddr_in6))
        delete reinterpret_cast<struct sockaddr_in6*>(ai.ai_addr);

```